spacy/training/example.pyx

```diff
@@ -150,27 +150,84 @@ cdef class Example:
                 self._y_sig = y_sig
                 return self._cached_alignment
 
-    def get_aligned(self, field, as_string=False):
-        """Return an aligned array for a token attribute."""
-        align = self.alignment.x2y
 
-        vocab = self.reference.vocab
-        gold_values = self.reference.to_array([field])
+    def _get_aligned_vectorized(self, align, gold_values):
+        # Fast path for Doc attributes/fields that are predominantly a single value,
+        # i.e., TAG, POS, MORPH.
+        x2y_single_toks = []
+        x2y_single_toks_i = []
+
+        x2y_multiple_toks = []
+        x2y_multiple_toks_i = []
+
+        # Gather indices of gold tokens aligned to the candidate tokens into two buckets.
+        #   Bucket 1: All tokens that have a one-to-one alignment.
+        #   Bucket 2: All tokens that have a one-to-many alignment.
+        for idx, token in enumerate(self.predicted):
+            aligned_gold_i = align[token.i]
+            if len(aligned_gold_i) == 1:
+                x2y_single_toks.append(aligned_gold_i.item())
+                x2y_single_toks_i.append(idx)
+            else:
+                x2y_multiple_toks.append(aligned_gold_i)
+                x2y_multiple_toks_i.append(idx)
+
+        # Map elements of the first bucket directly to the output array.
+        output = numpy.full(len(self.predicted), None)
+        output[x2y_single_toks_i] = gold_values[x2y_single_toks].squeeze()
+
+        # Collapse many-to-one alignments into one-to-one alignments if they
+        # share the same value. Map to None in all other cases.
+        for i in range(len(x2y_multiple_toks)):
+            aligned_gold_values = gold_values[x2y_multiple_toks[i]]
+
+            # If all aligned tokens have the same value, use it.
+            if len(set(list(aligned_gold_values))) == 1:
+                x2y_multiple_toks[i] = aligned_gold_values[0].item()
+            else:
+                x2y_multiple_toks[i] = None
+
+        output[x2y_multiple_toks_i] = x2y_multiple_toks
+
+        return output.tolist()
+
+
+    def _get_aligned_non_vectorized(self, align, gold_values):
+        # Slower path for fields that return multiple values (resulting
+        # in ragged arrays that cannot be vectorized trivially).
         output = [None] * len(self.predicted)
+
         for token in self.predicted:
-            values = gold_values[align[token.i]]
-            values = values.ravel()
+            aligned_gold_i = align[token.i]
+            values_org = gold_values[aligned_gold_i]
+            values = values_org.ravel()
```

Comment:
I think I liked the original more

```suggestion
            values = gold_values[aligned_gold_i].revel()
```

Because then the intermediate value can be removed before the end of the scope. If the Python interpreter is smart enough, it could infer this because `values_org` is not referenced after this point, but I wouldn't be too confident about this.

Reply:
Oh, I think I renamed that variable to print it out in whilst debugging and then forgot to revert that change.